crates/autopilot/src/decoded_settlement.rs

```diff
@@ -226,69 +236,72 @@ impl DecodedSettlement {
                     flags: trade.8.into(),
                     executed_amount: trade.9,
                     signature: trade.10,
-                })
-                .collect(),
+                };
+                match trade.compute_uid(domain_separator, &tokens) {
+                    Ok(order_uid) => {
+                        trade.order_uid = order_uid;
+                        Some(trade)
+                    }
+                    Err(err) => {
+                        tracing::error!(
+                            ?err,
+                            ?trade,
+                            "failed to calculate order uid, we don't know which order this trade \
+                             belongs to"
+                        );
+                        None
+                    }
+                }
+            })
+            .collect();
+        Ok(Self {
+            tokens,
+            clearing_prices,
+            trades,
             interactions: interactions.map(|inner| inner.into_iter().map(Into::into).collect()),
             metadata: metadata.try_into().ok().map(Bytes),
         })
     }
 
     /// Returns the total surplus denominated in the native asset for the
     /// solution.
-    pub fn total_surplus(&self, external_prices: &ExternalPrices) -> U256 {
-        self.trades.iter().fold(0.into(), |acc, trade| {
-            acc + match surplus(trade, &self.tokens, &self.clearing_prices, external_prices) {
-                Some(surplus) => surplus,
-                None => {
-                    tracing::warn!("possible incomplete surplus calculation");
-                    0.into()
-                }
-            }
-        })
-    }
-
-    /// Returns fees for all trades.
-    pub fn all_fees(
+    pub fn total_surplus(
         &self,
         external_prices: &ExternalPrices,
-        domain_separator: &DomainSeparator,
-    ) -> Vec<Fees> {
+        surplus_eligible_orders: &HashSet<OrderUid>,
```

Comment:
This parameter probably deserves some explanation in the docs of this method. I think we should explain to a developer that stumbles across this code in 6 months why we need to have a concept of eligible orders...

nit:
```suggestion
        eligible_orders: &HashSet<OrderUid>,
```